doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -762,58 +683,287 @@ the conversion in hypotheses :n:`{+ @ident}`.
          pattern (0 = 0).
          fold not.
 
-   .. tacv:: fold {+ @term}
+   .. example:: Use :tacn:`fold` to reverse unfolding of `fold_right`
 
-      Equivalent to :n:`fold @term ; ... ; fold @term`.
+      .. coqtop:: none
 
-.. tacn:: pattern @term
-   :name: pattern
+         Require Import Coq.Lists.List.
+         Local Open Scope list_scope.
 
-   This command applies to any goal. The argument :n:`@term` must be a free
-   subterm of the current goal. The command pattern performs :math:`\beta`-expansion
-   (the inverse of :math:`\beta`-reduction) of the current goal (say :g:`T`) by
+      .. coqtop:: all abort
 
-   + replacing all occurrences of :n:`@term` in :g:`T` with a fresh variable
-   + abstracting this variable
-   + applying the abstracted goal to :n:`@term`
+         Goal forall x xs, fold_right and True (x::xs).
+         red.
+         fold (fold_right and True).
 
-   For instance, if the current goal :g:`T` is expressible as
-   :math:`\varphi`:g:`(t)` where the notation captures all the instances of :g:`t`
-   in :math:`\varphi`:g:`(t)`, then :n:`pattern t` transforms it into
-   :g:`(fun x:A =>` :math:`\varphi`:g:`(x)) t`. This tactic can be used, for
-   instance, when the tactic ``apply`` fails on matching.
+.. tacn:: pattern {+, @pattern_occs } {? @occurrences }
 
-.. tacv:: pattern @term at {+ @natural}
+   Performs beta-expansion (the inverse of :term:`beta-reduction`) for the
+   selected hypotheses and/or goals.
+   The :n:`@one_term`\s in :n:`@pattern_occs` must be free subterms in the selected items.
+   The expansion is done for each selected item :g:`T`
+   for a set of :n:`@one_term`\s in the :n:`@pattern_occs` by:
 
-   Only the occurrences :n:`{+ @natural}` of :n:`@term` are considered for
-   :math:`\beta`-expansion. Occurrences are located from left to right.
+   + replacing all selected occurrences of the :n:`@one_term`\s in :g:`T` with fresh variables
+   + abstracting these variables
+   + applying the abstracted goal to the :n:`@one_term`\s
 
-.. tacv:: pattern @term at - {+ @natural}
+   For instance, if the current goal :g:`T` is expressible as :n:`φ(t__1 … t__n)`
+   where the notation captures all the instances of the :n:`t__i` in φ, then :tacn:`pattern`
+   :n:`t__1, …, t__n` generates the equivalent goal
+   :n:`(fun (x__1:A__1 … (X__n:A__n) => φ(x__1 … x__n)) t__1 … t__n`.
+   If :n:`t__i` occurs in one of the generated types :n:`A__j`
+   (for `j > i`),
+   occurrences will also be considered and possibly abstracted.
 
-   All occurrences except the occurrences of indexes :n:`{+ @natural }`
-   of :n:`@term` are considered for :math:`\beta`-expansion. Occurrences are located from
-   left to right.
+   This tactic can be used, for instance, when the tactic :tacn:`apply` fails on matching.
 
-.. tacv:: pattern {+, @term}
+vm_compute and native_compute
```

Comment:
Perhaps the title could be "Fast reduction machines"?

Reply:
How about "Fast reduction tactics: vm_compute and native_compute"?

![image](https://user-images.githubusercontent.com/1253341/109711780-1f728400-7b54-11eb-9563-20d887daa0e4.png)

Reply:
Yes, sounds good.